planner/core/optimizer.go

```diff
@@ -517,62 +517,121 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
 
 // eliminateUnionScanAndLock set lock property for PointGet and BatchPointGet and eliminates UnionScan and Lock.
 func eliminateUnionScanAndLock(sctx sessionctx.Context, p PhysicalPlan) PhysicalPlan {
-	var pointGet *PointGetPlan
-	var batchPointGet *BatchPointGetPlan
-	var physLock *PhysicalLock
-	var unionScan *PhysicalUnionScan
-	iteratePhysicalPlan(p, func(p PhysicalPlan) bool {
-		if len(p.Children()) > 1 {
-			return false
-		}
-		switch x := p.(type) {
-		case *PointGetPlan:
-			pointGet = x
-		case *BatchPointGetPlan:
-			batchPointGet = x
-		case *PhysicalLock:
-			physLock = x
-		case *PhysicalUnionScan:
-			unionScan = x
-		}
-		return true
-	})
-	if pointGet == nil && batchPointGet == nil {
+	type Tuple struct {
+		// Point get and batch point get cannot be not nil simultaneously
+		pointGet      *PointGetPlan
+		batchPointGet *BatchPointGetPlan
+
+		physicalLock *PhysicalLock
+	}
+
+	tuples := make([]*Tuple, 0)
+
+	// physicalLockMap is a map for mapping PhysicalLock to PhysicalUnionScan. We need this map because when
+	// getLockWaitTime(sctx, physicalLock.Lock) returns false, we delete the relevant PhysicalUnionScan (if it exists)
+	// in the unionScanSet to avoid elimination.
+	physicalLockMap := make(map[PhysicalPlan]PhysicalPlan)
+	unionScanSet := make(map[PhysicalPlan]interface{})
+
+	iteratePhysicalPlan(p, nil, nil,
+		func(p PhysicalPlan, physicalLock *PhysicalLock, unionScan *PhysicalUnionScan) (*PhysicalLock, *PhysicalUnionScan) {
+			// PhysicalUnionScan and PhysicalLock can only relate to one point get/batch point get.
+			// So, if the plan has multiple children, we should reset physicalLockMap and unionScanSet
+			if len(p.Children()) > 1 {
+				physicalLockMap = make(map[PhysicalPlan]PhysicalPlan)
+				unionScanSet = make(map[PhysicalPlan]interface{})
```

Comment:
So `iteratePhysicalPlan` performs a dfs and is there such a risk that we generate `physicalLockMap` and `unionScanSet` in one branch but reset them in another?

Reply:
The intention of this is that we don't want one physical lock or union scan relates to mutiple point get/batch point get.

The current implementation has some problems still, I refactor it later.